criu/pie/restorer.c

```diff
@@ -200,10 +205,19 @@ static int restore_creds(struct thread_creds_args *args, int procfd, int lsm_typ
 	 * Setup supplementary group IDs early.
 	 */
 	if (args->groups) {
-		ret = sys_setgroups(ce->n_groups, args->groups);
-		if (ret) {
-			pr_err("Can't setup supplementary group IDs: %d\n", ret);
-			return -1;
+		/*
+		 * We may be in an unprivileged user namespace where setgroups
+		 * is disabled.  If the current list of groups is already what
+		 * we want, skip the call to setgroups.
+		 */
+		unsigned int gids[MAX_GETGROUPS_CHECKED];
+		int n = sys_getgroups(MAX_GETGROUPS_CHECKED, gids);
+		if (n != ce->n_groups || memcmp(gids, args->groups, n * sizeof(*gids))) {
```

Comment:
do we have any guarantees that the groups list is sorted?

Reply:
According to a comment in `groups_match()` in 'cr-restore.c', `setgroups()` does sort the gids. I can't see that information in the manpage, though.